kafka_consumer/check.py

```diff
@@ -29,35 +40,161 @@ def __init__(self, name, init_config, agentConfig, instances=None):
+        """
+        Fetch Consumer Group offsets from Zookeeper.
+
+        Also fetch consumer_groups, topics, and partitions if not
+        already specified in consumer_groups.
+
+        :param dict consumer_groups: The consumer groups, topics, and partitions
+            that you want to fetch offsets for. If consumer_groups is None, will
+            fetch offsets for all consumer_groups. For examples of what this
+            dict can look like, see _validate_consumer_groups().
+        """
+        zk_consumer_offsets = {}
 
         # Construct the Zookeeper path pattern
-        zk_prefix = instance.get('zk_prefix', '')
-        zk_path_tmpl = zk_prefix + '/consumers/%s/offsets/%s/%s'
+        # /consumers/[groupId]/offsets/[topic]/[partitionId]
+        zk_path_consumer = zk_prefix + '/consumers/'  # /consumers/
+        zk_path_topic_tmpl = zk_path_consumer + '%s/offsets/'  # /consumers/[groupID]/offsets/
+        zk_path_partition_tmpl = zk_path_topic_tmpl + '%s/'  # /consumers/[groupID]/offsets/[topic]/
 
-        # Connect to Zookeeper
-        zk_conn = KazooClient(zk_connect_str, timeout=self.zk_timeout)
+        zk_conn = KazooClient(zk_hosts_ports, timeout=self.zk_timeout)
         zk_conn.start()
-
         try:
-            # Query Zookeeper for consumer offsets
-            consumer_offsets = {}
-            topics = defaultdict(set)
-            for consumer_group, topic_partitions in consumer_groups.iteritems():
-                for topic, partitions in topic_partitions.iteritems():
-                    # Remember the topic partitions that we've see so that we can
-                    # look up their broker offsets later
-                    topics[topic].update(set(partitions))
+            # Specifying consumer_groups is optional, if they don't exist, then fetch from ZK
+            if consumer_groups is None:
+                consumer_groups = {consumer_group: None for consumer_group in
+                    self._get_zk_path_children(zk_conn, zk_path_consumer, 'consumer groups')}
+
+            for consumer_group, topics in consumer_groups.iteritems():
+                # Specifying topics is optional, if they don't exist, then fetch from ZK
+                if topics is None:
+                    zk_path_topics = zk_path_topic_tmpl % (consumer_group)
+                    topics = {topic: None for topic in
+                        self._get_zk_path_children(zk_conn, zk_path_topics, 'topics')}
+
+                for topic, partitions in topics.iteritems():
+                    # Specifying partitions is optional, if they don't exist, then fetch from ZK
+                    if partitions is None:
+                        zk_path_partitions = zk_path_partition_tmpl % (consumer_group, topic)
+                        # Zookeeper returns the partition IDs as strings because
+                        # they are extracted from the node path
+                        partitions = [int(x) for x in self._get_zk_path_children(
```

Comment:
Although we can be certain `partitions` here will have no repeated elements, we can't say the same about the list passed through by the user (in the YAML). So we should probably continue using a set to avoid unnecessary duplicated calls. Alternatively, we can maybe sanitize `consumer_groups` in `check()` before passing it on? Your call.